app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -0,0 +1,102 @@
+    ) do
+
+      def call
+        parse_xms_mirid
+        # compare xms_mirid with what is defined in annotations
+        validate_application_identity
+      end
+
+      private
+
+      # validate integrity of annotations
+      def application_identity
+        @application_identity ||= @application_identity_class.new(
+            role_annotations: role_annotations,
+            service_id: @service_id
+        )
+      end
+
+      def role_annotations
+        role.annotations
+      end
+
+      # convert host ID to full host ID, cucumber:host:somepolicy
+      # TODO For a Conjur user, an incorrect full id will be returned
+      def role_id
+        @role_id ||= @role_class.roleid_from_username(@account, @username)
+      end
+
+      def role
+        @role ||= @resource_class[role_id]
+        raise SecurityErr::RoleNotFound(role_id) if @username.nil?
+        return @role
+      end
+
+      # TODO add parsing format to check xms mirid is in proper format
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # subscription, resource group, and provider identity needed for authorization. xms_mirid is one of the fields in
+      # the JWT token. This function will extract the relevant information from xms_mirid claim and populate a representative
+      # hash with the appropriate fields.
+      def parse_xms_mirid
+        # xms_mirid claim starts with an extra '/' so we will be ignoring it for parsing purposes
+        # ex: /subscription/<subscription_id> -> subscription/<subscription_id>
+        _, *field_split = @xms_mirid.split('/')
+        @xms_mirid_hash = Hash[field_split.each_slice(2).to_a]
+
+        @token_identity = {
+            subscription_id: @xms_mirid_hash["subscriptions"],
+            resource_group: @xms_mirid_hash["resourcegroups"]
+        }
+
+        # determines which Azure assigned identity is provided in annotations
+        # user-assigned-identity:
+        #   - validates that the correct provider (Microsoft.ManagedIdentity) for a user identity is defined in the
+        #     claim. If so, the 'user_assigned_identity' attribute will be added to the hash with the corresponding
+        #     value from xms_mirid claim
+        # system-assigned-identity:
+        #   - validates that the correct provider (Microsoft.Compute) for a system identity is defined in the
+        #     claim and its resource is one that we support. At current, we only support a virtualMachines resource.
+        #     If these conditions are met, a 'system_assigned_identity' attribute will be added to the hash with its
+        #     value being the oid field in the JWT token.
+        if @xms_mirid_hash["providers"] == "Microsoft.ManagedIdentity"
+          @logger.debug(Log::AuthnAzure::ExtractingIdentityForAuthentication.new(@xms_mirid_hash["providers"] << "/" << @xms_mirid_hash.keys[-1]))
```

Comment:
`@xms_mirid_hash["providers"] << "/" << @xms_mirid_hash.keys[-1]` -> `"#{@xms_mirid_hash["providers"]}/#{@xms_mirid_hash.keys[-1]}

Reply:
intentional `"#` w.o a closing "?

Reply:
updating to `"#{@xms_mirid_hash["providers"]}/#{@xms_mirid_hash.keys[-1]}"`